Library/Homebrew/test/utils/shell_spec.rb

```diff
@@ -0,0 +1,68 @@
+require "utils/shell"
+
+describe Utils::Shell do
+  describe "::path_to_shell" do
+    it "supports a raw command name" do
+      expect(described_class.path_to_shell("bash")).to eq(:bash)
+    end
+
+    it "supports full paths" do
+      expect(described_class.path_to_shell("/bin/bash")).to eq(:bash)
+    end
+
+    it "supports versions" do
+      expect(described_class.path_to_shell("zsh-5.2")).to eq(:zsh)
+    end
+
+    it "strips newlines" do
+      expect(described_class.path_to_shell("zsh-5.2\n")).to eq(:zsh)
+    end
+
+    it "returns nil when input is invalid" do
+      expect(described_class.path_to_shell("")).to be nil
+      expect(described_class.path_to_shell("@@@@@@")).to be nil
+      expect(described_class.path_to_shell("invalid_shell-4.2")).to be nil
+    end
+  end
+
+  specify "::sh_quote" do
+    expect(described_class.sh_quote("")).to eq("''")
+    expect(described_class.sh_quote("\\")).to eq("\\\\")
+    expect(described_class.sh_quote("\n")).to eq("'\n'")
+    expect(described_class.sh_quote("$")).to eq("\\$")
+    expect(described_class.sh_quote("word")).to eq("word")
+  end
+
+  specify "::csh_quote" do
+    expect(described_class.csh_quote("")).to eq("''")
+    expect(described_class.csh_quote("\\")).to eq("\\\\")
+    # note this test is different than for sh
+    expect(described_class.csh_quote("\n")).to eq("'\\\n'")
+    expect(described_class.csh_quote("$")).to eq("\\$")
+    expect(described_class.csh_quote("word")).to eq("word")
+  end
+
+  describe "::prepend_path_in_shell_profile" do
+    subject { described_class.prepend_path_in_shell_profile(path) }
+    let(:path) { "/path" }
+
+    before(:each) do
+      ENV["SHELL"] = shell
+    end
+
+    context "tcsh" do
```

Comment:
Splitting all these into contexts with a single example inside adds quite a bit of indirection, and I'm not sure what benefit it provides over just having tests like this:

```ruby
it "supports tcsh" do
  expect(described_class.prepend_path_in_shell_profile("/bin/tcsh")).
    to_start_with("echo 'setenv PATH /path")
end
```